### PR TITLE
Add MCP Registry UI foundation: types, API client, page shell, and sidebar entry

### PR DESCRIPTION
**Pull request by DaoDaoNoCode**
### Related Issues/PRs

Relates to #23695

### What changes are proposed in this pull request?

This is the first in a series of PRs adding the MCP Registry UI, building on the backend REST API from #23695. It sets up the foundational pieces so subsequent PRs can add list views, detail pages, and access binding management incrementally.
Screenshot 2026-06-23 at 7 12 07 PM
Screenshot 2026-06-23 at 7 12 35 PM

**TypeScript types** (`types.ts`)
- Entity types for `MCPServer`, `MCPServerVersion`, `MCPAccessBinding`, `MCPServerAlias`, `MCPIcon`, and the `ServerJSONPayload` spec types — all matching the Pydantic models from the backend.
- Request/response interfaces for all CRUD operations and search endpoints.

**API client** (`api.ts`)
- `MCPRegistryApi` object covering all 24 REST endpoints under `/ajax-api/3.0/mlflow/mcp-servers`.
- Proper `encodeURIComponent` for server names containing `/` (e.g., `io.github.example/my-server`).

**Routing and navigation**
- `route-defs.ts` and `routes.ts` with lazy-loaded page component and document title handle.
- Sidebar entry under "MCP Registry" gated behind `shouldEnableMCPRegistryUI`.

**Page shell** (`MCPRegistryPage.tsx`)
- Two tabs: "Access Bindings" (default) and "Servers".
- Search filter input with debounce, grid/list view toggle.
- Empty states for when no servers exist.
- Placeholder "Create MCP server" button (disabled for now — creation comes in a later PR).

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

Verified the page loads correctly with the sidebar entry, tabs switch, search filter works, and empty states render. Type-checked with `tsc --noEmit`, linted, and prettier-checked.

### Does this PR require documentation update?

- [x] No.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.
- [ ] Yes. Please link the corresponding PR or explain how you plan to update it.

#### Is this a user-facing change?

- [x] Yes. Adds the MCP Registry page shell with sidebar navigation — the starting point for the MCP server registry UI.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

**Comment by DaoDaoNoCode** [comment]
@mprahl @nananosirova Can you review the changes?

File: mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+      <Spacer shrinks={false} />
+      <div css={{ flex: 1, display: 'flex', flexDirection: 'column', overflow: 'hidden' }}>
+        <SegmentedControlGroup
+          name="mcp-registry-tabs"
+          value={activeTab}
+          onChange={handleTabChange}
+          componentId="mlflow.mcp_registry.tabs"
+        >
+          <SegmentedControlButton value="servers">
+            <FormattedMessage defaultMessage="Servers" description="MCP Registry servers tab label" />
```
Comment (nananosirova) [inline mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx:109]: Should access bindings show up first in the toggle?

File: mlflow/server/js/src/mcp-registry/types.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+export interface ServerJSONPayload {
+  name: string;
+  version: string;
+  title?: string;
+  description?: string;
+  packages?: ServerJSONPackage[];
+  remotes?: ServerJSONRemote[];
+  repository?: string;
```
Comment (nananosirova) [inline mlflow/server/js/src/mcp-registry/types.ts]: `repository` is typed as `string` here but in the pydantic model it's `ServerJSONRepositoryPayload` object:  https://github.com/mlflow/mlflow/blob/mcp-registry/mlflow/server/mcp_server_api.py#L67

File: mlflow/server/js/src/mcp-registry/types.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  inputSchema?: Record<string, unknown>;
+  outputSchema?: Record<string, unknown>;
+  annotations?: Record<string, unknown>;
+  icons?: MCPIcon[];
+  execution?: Record<string, unknown>;
+}
+
+export interface MCPIcon {
+  src: string;
+  sizes?: string;
```
Comment (nananosirova) [inline mlflow/server/js/src/mcp-registry/types.ts]: `sizes` is typed as `string` but the pydantic model has `list[str]`:https://github.com/mlflow/mlflow/blob/mcp-registry/mlflow/server/mcp_server_api.py#L32

Should this be `string[]`?

File: mlflow/server/js/src/mcp-registry/types.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+  name: string;
+  description?: string;
+  isRequired?: boolean;
+  isSecret?: boolean;
+}
+
+export interface ServerJSONPackage {
+  registryType: string;
+  identifier: string;
+  transport: string;
```
Comment (nananosirova) [inline mlflow/server/js/src/mcp-registry/types.ts:99]: Maybe we could tighten the type for `transport` since we're parsing `server.json` later. Each `type` (`stdio | streamable-http | sse`) variant seems to have a different shape: 
```typescript
type StdioTransport = { type: 'stdio'; command: string; args?: string[] };
type StreamableHttpTransport = { type: 'streamable-http'; url: string };
type SseTransport = { type: 'sse'; url: string };
type ServerJSONTransport = StdioTransport | StreamableHttpTransport | SseTransport;
```
The pydantic model just sets it to `any` at the moment: https://github.com/mlflow/mlflow/blob/mcp-registry/mlflow/server/mcp_server_api.py#L86

File: mlflow/server/js/src/mcp-registry/types.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+  theme?: string;
+}
+
+export interface MCPAccessBindingSummary {
+  binding_id: number;
+  endpoint_url: string;
+  transport_type: MCPRemoteTransportType;
+  server_version?: string;
+  server_alias?: string;
+  resolved_version?: MCPServerVersion;
```
Comment (nananosirova) [inline mlflow/server/js/src/mcp-registry/types.ts:32]: Should `server_name : string` also be added? 
https://github.com/mlflow/mlflow/blob/mcp-registry/mlflow/server/mcp_server_api.py#L177

File: mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+              </TableRow>
+            </Table>
+          </>
+        )}
+      </div>
+    </ScrollablePageWrapper>
+  );
+};
+
+export default withErrorBoundary(ErrorUtils.mlflowServices.EXPERIMENTS, MCPRegistryPage);
```
Comment (nananosirova) [inline mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx]: I see using `EXPERIMENTS` for the error boundary seems pretty common, but should we just define our own here?

File: mlflow/store/tracking/dbmodels/models.py [hunk 7]
```diff
@@ -24,7 +24,14 @@
...
 from sqlalchemy.inspection import inspect
-from sqlalchemy.orm import aliased, backref, query_expression, relationship, validates, with_expression
+from sqlalchemy.orm import (
+    aliased,
+    backref,
+    query_expression,
+    relationship,
+    validates,
+    with_expression,
+)
```
Comment (nananosirova) [inline mlflow/store/tracking/dbmodels/models.py:33]: nit: unnecessary change

**Comment by nananosirova** [comment]
Maybe we could also add some API client tests? And page render, tab switch and url update, empty state tests?

File: mlflow/server/js/src/mcp-registry/types.ts [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+  name: string;
+  description?: string;
+  icons?: MCPIcon[];
+}
+
+export interface UpdateMCPServerRequest {
+  display_name?: string | null;
+  description?: string | null;
+  icons?: MCPIcon[] | null;
+  latest_version?: string | null;
```
Comment (nananosirova) [inline mlflow/server/js/src/mcp-registry/types.ts]: `latest_version` has also recently been made read-only: https://github.com/mlflow/mlflow/commit/f0a26e8556c3c70ba9772c85c9992930b27ae59e
It would have to be removed from `UpdateMCPServerRequest`

File: mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+              </TableFilterLayout>
+            </div>
+            <Table
+              scrollable
+              empty={
+                <div css={emptyCenterStyles}>
+                  <Empty
+                    title={
+                      <FormattedMessage
+                        defaultMessage="Create endpoint"
```
Comment (nananosirova) [inline mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx]: The empty state for Access Bindings says "Create endpoint" / "Create and manage direct access endpoints for your MCP servers.". Should this say "Create access binding" instead to be consistent with the naming?

**Comment by nananosirova** [comment]
Here's an updated auto-generated OpenAPI spec: https://editor.swagger.io/?url=https://gist.githubusercontent.com/nananosirova/7ab184128c67fc3e633b09f609aa934b/raw/mcp-openapi-latest.json

File: mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+            <div css={{ paddingTop: theme.spacing.md }}>
+              <TableFilterLayout>
+                <TableFilterInput
+                  placeholder={intl.formatMessage({
+                    defaultMessage: 'Search access bindings',
+                    description: 'Placeholder for MCP access bindings search filter input',
+                  })}
+                  componentId="mlflow.mcp_registry.bindings.search"
+                  value={searchFilter}
+                  onChange={(e) => setSearchFilter(e.target.value)}
```
Comment (Gkrumbach07) [inline mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx:226]: we should set up debounce here and in the other filter change too.

File: mlflow/server/js/src/mcp-registry/types.ts (hunk 2 above)
Comment (DaoDaoNoCode) [reply mlflow/server/js/src/mcp-registry/types.ts]: Fixed — added `ServerJSONRepository` interface with `url`, `source`, `id` fields matching the backend model.

File: mlflow/server/js/src/mcp-registry/types.ts (hunk 3 above)
Comment (DaoDaoNoCode) [reply mlflow/server/js/src/mcp-registry/types.ts]: Fixed — changed to `string[]`.

File: mlflow/server/js/src/mcp-registry/types.ts (hunk 4 above)
Comment (DaoDaoNoCode) [reply mlflow/server/js/src/mcp-registry/types.ts:99]: Good idea. Since the backend currently types this as `any` and the spec is still evolving (e.g. new transport types may be added), I kept it loose for now with `type: string` + `url?: string`. Can tighten once the spec stabilizes. Open to revisiting in a follow-up.

File: mlflow/server/js/src/mcp-registry/types.ts (hunk 5 above)
Comment (DaoDaoNoCode) [reply mlflow/server/js/src/mcp-registry/types.ts:32]: Added.

File: mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx (hunk 6 above)
Comment (DaoDaoNoCode) [reply mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx]: Good call — added `MCP_REGISTRY` to `ErrorUtils.mlflowServices` and updated the page to use it.

File: mlflow/store/tracking/dbmodels/models.py (hunk 7 above)
Comment (DaoDaoNoCode) [reply mlflow/store/tracking/dbmodels/models.py:33]: Reverted — removed from our diff. The lint issue is pre-existing on the base branch.

File: mlflow/server/js/src/mcp-registry/types.ts (hunk 8 above)
Comment (DaoDaoNoCode) [reply mlflow/server/js/src/mcp-registry/types.ts]: Removed `latest_version` from `UpdateMCPServerRequest`. The field is now read-only on the server.

File: mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx (hunk 9 above)
Comment (DaoDaoNoCode) [reply mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx]: Fixed — changed to "Create access binding" / "Create and manage access bindings for your MCP servers."

File: mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx (hunk 10 above)
Comment (DaoDaoNoCode) [reply mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx:226]: Good call — debounce will be added in the next PR when we wire up data fetching hooks. This PR is just the page shell with local state, so no API calls are triggered by the filter yet.

**Comment by mprahl** [comment]
@DaoDaoNoCode could you please rebase?

File: mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx (hunk 1 above)
Comment (DaoDaoNoCode) [reply mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx:109]: I will add it in the follow up PRs, this is a shell page, so we keep it as it is in this PR.

**Comment by DaoDaoNoCode** [comment]
Added tests — 6 test cases covering page render, both tabs, tab switching, search input placeholder, and empty states for both tabs. Also added `test-utils.ts` with mock factories for reuse in follow-up PRs. Rebased as well.

File: mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+                </TableFilterLayout>
+              </div>
+              <SegmentedControlGroup
+                name="mcp-registry-view-mode"
+                value={viewMode}
+                onChange={(e) => setViewMode(e.target.value as ViewMode)}
+                componentId="mlflow.mcp_registry.view_toggle"
+              >
+                <SegmentedControlButton value="list" icon={<ListIcon />} />
+                <SegmentedControlButton value="grid" icon={<GridIcon />} />
```
Comment (mprahl) [inline mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx]: Could we follow the existing MLflow view-toggle pattern here and add labels for these icon-only segmented buttons, ideally via tooltips and hidden text? Other places in the UI do that for similar list/table/chart toggles, which makes the control more discoverable and a bit friendlier from an accessibility standpoint.

🤖 Generated with Claude

Comment (DaoDaoNoCode) [reply mlflow/server/js/src/mcp-registry/pages/MCPRegistryPage.tsx]: Added `aria-label` to both view toggle buttons ("List view" / "Grid view") for accessibility.

File: mlflow/server/js/src/mcp-registry/types.ts [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+
+// Request types
+
+export interface CreateMCPServerRequest {
+  name: string;
+  description?: string;
+  icons?: MCPIcon[];
+}
+
+export interface UpdateMCPServerRequest {
```
Comment (mprahl) [inline mlflow/server/js/src/mcp-registry/types.ts:140]: **🤖 AI Review**

`UpdateMCPServerRequest` is missing `latest_version`, even though the backend PATCH contract accepts it in `mlflow/server/mcp_server_api.py` and `tests/server/test_mcp_server_api.py` already exercises that path. Since `MCPRegistryApi.updateMCPServer()` uses this type directly, typed callers currently can't set or clear the pinned latest version without bypassing TypeScript. Please add `latest_version?: string | null` here so the frontend contract stays aligned with the API.

File: mlflow/server/js/src/mcp-registry/types.ts [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+  [key: string]: unknown;
+}
+
+export interface ServerJSONRemote {
+  type: string;
+  url: string;
+  [key: string]: unknown;
+}
+
+export interface ServerJSONRepository {
```
Comment (mprahl) [inline mlflow/server/js/src/mcp-registry/types.ts:112]: **🤖 AI Review**

`ServerJSONRepository` doesn't include `subfolder`, but the backend request model accepts it in `mlflow/server/mcp_server_api.py` and `tests/server/test_mcp_server_api.py` already covers a repository payload that uses it. As written, valid repository-backed `server_json` payloads will fail the frontend type check. Please add `subfolder?: string` so this type matches the supported request shape.

**Review by mprahl — COMMENTED** [review]
**🤖 AI Review**

File: mlflow/server/js/src/mcp-registry/types.ts (hunk 12 above)
Comment (DaoDaoNoCode) [reply mlflow/server/js/src/mcp-registry/types.ts:140]: `latest_version` was intentionally removed from `UpdateMCPServerRequest` — it was recently made read-only on the backend (commit f0a26e8). The server now rejects it with a validator: `"latest_version is read-only; it is resolved automatically"`. This was also flagged and confirmed in Nana's review above.

File: mlflow/server/js/src/mcp-registry/types.ts (hunk 13 above)
Comment (DaoDaoNoCode) [reply mlflow/server/js/src/mcp-registry/types.ts:112]: Good catch — added `subfolder?: string` to `ServerJSONRepository`.

File: mlflow/server/js/src/common/components/MlflowSidebar.tsx [hunk 14]
```diff
@@ -236,6 +239,22 @@ export function MlflowSidebar({
...
+      ...(shouldShowGenAIFeatures(enableWorkflowBasedNavigation, workflowType) && !showNestedExperimentItems
+        ? [
+            {
+              key: 'mcp-registry',
+              icon: <WrenchIcon />,
+              linkProps: {
+                to: MCPRegistryRoutes.mcpRegistryPageRoute,
+                isActive: isMCPRegistryActive,
+                children: (
+                  <FormattedMessage defaultMessage="MCP Registry" description="Sidebar link for MCP Registry page" />
```
Comment (mprahl) [inline mlflow/server/js/src/common/components/MlflowSidebar.tsx]: To match "Model registry" casing.
```suggestion
                  <FormattedMessage defaultMessage="MCP registry" description="Sidebar link for MCP registry page" />
```

**Review by mprahl — CHANGES_REQUESTED** [review]
One small nit and then it looks good to me!

File: mlflow/server/js/src/common/components/MlflowSidebar.tsx (hunk 14 above)
Comment (DaoDaoNoCode) [reply mlflow/server/js/src/common/components/MlflowSidebar.tsx]: Fixed — changed to "MCP registry" to match "Model registry" casing. Also fixed prettier formatting on the aria-label lines.

**Review by mprahl — APPROVED** [review]
